test/e2e/reconciler/workspace/controller_test.go

```diff
@@ -23,26 +23,25 @@ import (
 	"testing"
 	"time"
 
-	"github.com/google/go-cmp/cmp"
-
 	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
-	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
 
 	"github.com/kcp-dev/kcp/pkg/apis/tenancy/helpers/conditions"
 	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
 	clientset "github.com/kcp-dev/kcp/pkg/client/clientset/versioned"
+	kcpexternalversions "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
 	"github.com/kcp-dev/kcp/test/e2e/framework"
 )
 
 func TestWorkspaceController(t *testing.T) {
 	type runningServer struct {
 		framework.RunningServer
-		client  clientset.Interface
-		watcher watch.Interface
+		client clientset.Interface
+		expect func(*tenancyv1alpha1.Workspace, func(workspace *tenancyv1alpha1.Workspace) error) error
```

Comment:
At first glance, this signature is confusing. Would like to brainstorm to see if we can make it easier to use

Reply:
Maybe
```go
type workspaceConditionFunc func(*Workspace) error
// ...
expect func(*Workspace, workspaceConditionFunc) error
```

Reply:
Yeah, I wanted opinions on this one... and to boot there's no great way to better encapsulate the server's ownership of this process. That is, the server knows how to divulge credentials, but we expect people to pass their own in, since getting from `*rest.Config` is yucky. Furthremore, we will likely always have the lister preamble in expectation funcs, but since those are typed I can't see a way to DRY that out simply.